File: core/youtube_classify.py
"""Route a YouTube video into a processing category.

A video reaches the pipeline two ways: as yt-dlp metadata (a dict from
``--dump-json`` / ``--flat-playlist``) when enumeration succeeds, or as a
yt-dlp stderr line when the download itself fails. Both need to collapse
to the same small bucket so the queue can decide whether to skip the
video, retry it later (live/premiere), or surface a friendly reason to
the user (members-only, age-restricted, region-locked).

This is a single pure function — no network, no subprocess. An
unrecognised error string is deliberately *not* classified here
(returns ``"ok"``) so the caller treats it as an ordinary failure.
"""

from __future__ import annotations

# Stable category strings. Keep these in sync with any caller that
# persists or branches on the result.
OK = "ok"
SHORT = "short"
LIVE = "live"
MEMBERS_ONLY = "members_only"
AGE_RESTRICTED = "age_restricted"
REGION_LOCKED = "region_locked"

ALL_CATEGORIES = (
    OK,
    SHORT,
    LIVE,
    MEMBERS_ONLY,
    AGE_RESTRICTED,
    REGION_LOCKED,
)

# Short, friendly, user-facing explanations. "ok" has no message.
_MESSAGES: dict[str, str] = {
    SHORT: "YouTube Short.",
    LIVE: "Live/premiere — will retry once it finishes.",
    MEMBERS_ONLY: "Members-only video — can't be downloaded.",
    AGE_RESTRICTED: "Age-restricted — needs sign-in, can't be downloaded.",
    REGION_LOCKED: "Blocked in this region.",
}


def classify_video(meta_or_error: dict | str) -> tuple[str, str]:
    """Return ``(category, message)`` for a YouTube video.

    ``category`` is one of:
      ``"ok"``             — normal downloadable video, ``message`` ``""``
      ``"short"``          — a YouTube Short
      ``"live"``           — live stream / premiere not finished (retry later)
      ``"members_only"``   — channel-members-only content
      ``"age_restricted"`` — requires sign-in / age confirmation
      ``"region_locked"``  — blocked in the current region

    ``meta_or_error`` is either yt-dlp metadata (a dict) or a yt-dlp
    stderr/error string. ``message`` is a short user-facing explanation
    (``""`` for ``"ok"``).
    """
    if isinstance(meta_or_error, dict):
        return _classify_meta(meta_or_error)
    return _classify_error(str(meta_or_error))


def _result(category: str) -> tuple[str, str]:
    return category, _MESSAGES.get(category, "")


def _classify_meta(meta: dict) -> tuple[str, str]:
    """Inspect yt-dlp metadata fields. Tolerant of missing keys."""
    # Live / premiere first — these are retry-later, not skip.
    live_status = meta.get("live_status")
    if live_status in ("is_live", "is_upcoming", "post_live") or meta.get("is_live"):
        return _result(LIVE)

    # Short — either an explicit /shorts/ URL or a <=60s duration. A
    # missing/None duration is NOT a short.
    url = meta.get("url") or meta.get("webpage_url") or ""
    if "/shorts/" in url:
        return _result(SHORT)
    duration = meta.get("duration")
    if duration is not None and duration <= 60:
        return _result(SHORT)

    # Availability / age gates.
    availability = meta.get("availability")
    if availability == "subscriber_only":
        return _result(MEMBERS_ONLY)
    age_limit = meta.get("age_limit") or 0
    if availability == "needs_auth" or age_limit >= 18:
        return _result(AGE_RESTRICTED)

    return OK, ""


# Known yt-dlp error phrases, matched case-insensitively as substrings.
# Ordered by priority: a members-only gate is reported before the more
# generic age/region/live phrases.
_ERROR_PHRASES: tuple[tuple[str, str], ...] = (
    ("join this channel to get access", MEMBERS_ONLY),
    ("members-only", MEMBERS_ONLY),
    ("members only", MEMBERS_ONLY),
    ("sign in to confirm your age", AGE_RESTRICTED),
    ("confirm your age", AGE_RESTRICTED),
    ("age-restricted", AGE_RESTRICTED),
    ("inappropriate for some users", AGE_RESTRICTED),
    ("not made this video available in your country", REGION_LOCKED),
    ("not available in your country", REGION_LOCKED),
    ("in your country", REGION_LOCKED),
    ("geo restrict", REGION_LOCKED),
    ("this live event will begin", LIVE),
    ("live event will begin", LIVE),
    ("premiere will begin", LIVE),
    ("premieres in", LIVE),
    ("is live and", LIVE),
)


def _classify_error(text: str) -> tuple[str, str]:
    """Match a yt-dlp stderr line against known error phrases."""
    lowered = text.lower()
    for phrase, category in _ERROR_PHRASES:
        if phrase in lowered:
            return _result(category)
    return OK, ""

File: Sources/VocatecaCore/Sources/YouTubeClassify.swift
import Foundation

// MARK: - YouTubeClassify

/// Oracle-locked port of `core/youtube_classify.py`.
///
/// Routes a YouTube video into a processing category from either yt-dlp metadata
/// (a dict) or a yt-dlp stderr error string. Pure function — no network, no I/O.
///
/// Do NOT change these algorithms without regenerating the golden fixtures and
/// running `swift test --filter OracleYouTubeTests`.
public enum YouTubeClassify {

    // MARK: - Category string constants

    /// Normal downloadable video (no message).
    public static let ok             = "ok"
    /// A YouTube Short (<= 60 s or `/shorts/` URL).
    public static let short          = "short"
    /// Live stream or premiere that has not finished (retry later).
    public static let live           = "live"
    /// Channel-members-only content.
    public static let membersOnly    = "members_only"
    /// Age-restricted video requiring sign-in.
    public static let ageRestricted  = "age_restricted"
    /// Blocked in the current region.
    public static let regionLocked   = "region_locked"

    // MARK: - User-facing messages (match Python _MESSAGES exactly)

    private static let messages: [String: String] = [
        short:         "YouTube Short.",
        live:          "Live/premiere \u{2014} will retry once it finishes.",
        membersOnly:   "Members-only video \u{2014} can't be downloaded.",
        ageRestricted: "Age-restricted \u{2014} needs sign-in, can't be downloaded.",
        regionLocked:  "Blocked in this region.",
    ]

    // MARK: - classify(meta:)

    /// Classify a video from yt-dlp metadata (dict).
    ///
    /// Port of `_classify_meta(meta)` from `core/youtube_classify.py`.
    ///
    /// Priority order:
    /// 1. Live/premiere (`live_status` in `{"is_live","is_upcoming","post_live"}` or `is_live` truthy)
    /// 2. Short (`/shorts/` URL or `duration <= 60`, None/missing duration is NOT a short)
    /// 3. `availability == "subscriber_only"` → members_only
    /// 4. `availability == "needs_auth"` or `age_limit >= 18` → age_restricted
    /// 5. Otherwise: ok
    ///
    /// - Parameter meta: The yt-dlp metadata dict encoded as `[String: JSONValue]`.
    /// - Returns: `(category, message)` tuple; message is `""` for `"ok"`.
    public static func classify(meta: [String: JSONValue]) -> (category: String, message: String) {
        // 1. Live / premiere
        let liveStatus = stringValue(meta["live_status"])
        if liveStatus == "is_live" || liveStatus == "is_upcoming" || liveStatus == "post_live" {
            return result(live)
        }
        if isTruthy(meta["is_live"]) {
            return result(live)
        }

        // 2. Short — /shorts/ URL or duration <= 60. None/missing is NOT a short.
        let url = stringValue(meta["url"]) ?? stringValue(meta["webpage_url"]) ?? ""
        if url.contains("/shorts/") {
            return result(short)
        }
        if let dur = numberValue(meta["duration"]) {
            if dur <= 60 {
                return result(short)
            }
        }

        // 3. Availability / age gates
        let availability = stringValue(meta["availability"])
        if availability == "subscriber_only" {
            return result(membersOnly)
        }
        let ageLimit = intValue(meta["age_limit"]) ?? 0
        if availability == "needs_auth" || ageLimit >= 18 {
            return result(ageRestricted)
        }

        return (ok, "")
    }

    // MARK: - classify(errorText:)

    /// Classify a video from a yt-dlp stderr/error string.
    ///
    /// Port of `_classify_error(text)` from `core/youtube_classify.py`.
    /// Matches known phrases case-insensitively as substrings, in priority order.
    ///
    /// - Parameter errorText: A yt-dlp stderr line or error description.
    /// - Returns: `(category, message)` tuple; `("ok", "")` for unrecognised errors.
    public static func classify(errorText: String) -> (category: String, message: String) {
        let lowered = errorText.lowercased()
        for (phrase, category) in errorPhrases {
            if lowered.contains(phrase) {
                return result(category)
            }
        }
        return (ok, "")
    }

    // MARK: - Private helpers

    private static func result(_ category: String) -> (String, String) {
        (category, messages[category] ?? "")
    }

    private static func stringValue(_ val: JSONValue?) -> String? {
        guard case .string(let s) = val else { return nil }
        return s
    }

    private static func numberValue(_ val: JSONValue?) -> Double? {
        guard case .number(let n) = val else { return nil }
        return n
    }

    private static func intValue(_ val: JSONValue?) -> Int? {
        guard case .number(let n) = val else { return nil }
        return Int(n)
    }

    private static func isTruthy(_ val: JSONValue?) -> Bool {
        switch val {
        case .bool(let b): return b
        case .number(let n): return n != 0
        case .string(let s): return !s.isEmpty
        case .null, nil: return false
        case .array(let a): return !a.isEmpty
        case .object(let o): return !o.isEmpty
        }
    }

    // MARK: - Error phrases (ordered by priority, matches Python _ERROR_PHRASES)

    private static let errorPhrases: [(String, String)] = [
        ("join this channel to get access", membersOnly),
        ("members-only",                    membersOnly),
        ("members only",                    membersOnly),
        ("sign in to confirm your age",     ageRestricted),
        ("confirm your age",                ageRestricted),
        ("age-restricted",                  ageRestricted),
        ("inappropriate for some users",    ageRestricted),
        ("not made this video available in your country", regionLocked),
        ("not available in your country",   regionLocked),
        ("in your country",                 regionLocked),
        ("geo restrict",                    regionLocked),
        ("this live event will begin",      live),
        ("live event will begin",           live),
        ("premiere will begin",             live),
        ("premieres in",                    live),
        ("is live and",                     live),
    ]
}
